### DISCUSSION TITLE
Tips on upgrading an application to WireUI

### DISCUSSION THREAD

**Discussion by judgej** [wireui/wireui#596] [General]
I'm adding some notes here in case they are useful. They may be obvious to some, not so much to others, there may be better ways to handle this, and there may be corrections to be made. Let me know what you think.

Okay, this is the situation I am in. I have an application with a tall-stack tonne of hand-made Livewire components. The components are a mix of working, not-quite-working, and are hard to maintain. My aim is to switch largely over to WireUI. The problem is that the names of the WireUI components clash with many of the hand-made components. So, how to migrate?

The first step is a simple installation. Just [follow the installation steps](https://livewire-wireui.com/docs/get-started) and publish the `wireui` config file:

    php artisan vendor:publish --tag='wireui.config'

Next, edit the `config/wireuid.php` confit file. Comment out all the lines in the `components` element:

```php
    'components' => [
        // 'avatar' => [
        //     'class' => Components\Avatar::class,
        //     'alias' => 'avatar',
        // ],
        // 'icon' => [
        //     'class' => Components\Icon::class,
        //     'alias' => 'icon',
        // ],
        // ...
    ],
```

That disables all the WireUI components - just the exiting application or other package components are available, so the application should work as before.

Then change the alias of all the components to add a prefix to "namespace" the components. The alias is the name of the component element you will use in Livewire blade templates. I use "wireui." as a prefix, like this:

```php
    'components' => [
        // 'avatar' => [
        //     'class' => Components\Avatar::class,
        //     'alias' => 'wireui.avatar',
        // ],
        // 'icon' => [
        //     'class' => Components\Icon::class,
        //     'alias' => 'wireui.icon',
        // ],
        // ...
    ],
```

Now you can start to migrate your legacy components. Taking this one as an example:

    <x-input wire:model.lazy="page_number" />

I upgrade that to the WireUI component using the prefix:

    <x-wireui.input wire:model.lazy="page_number" />

That one is straight-forward, but others are a little more tricky. For example, the legacy componnent may use icons that WireUI does not support (which can be added, or you just use the Heroki alternatives), or additional attributes that need changing or classes that need adding. But the point is, by using the new names with the prefix, you can migrate them all one-by-one, and the application will continue to work throughout the process.

Now, remember the components were all commented out? As you tackle each type of component, you need to uncomment them. I personally an uncomenting just the components I need. Now, many components are nested. So if you use this:

    <x-wireui.badge icon="check" positive label="Positive" />

You will need to uncomment the `badge` component. You will also need to uncomment the `icon` component, sine that component is used to add the icon to the badge. You will find these as you go through and blade complains about not being able to find a component.

And that is hopefully useful to someone. It will certainly be useful to me, when I've forgotten all about this on another project and go searching for an answer. Have fun.

Update: I think even with fresh projects I would add these prefixes. It just pays to keep everything namespaced in some way to avoid clashes, especially if you are pulling in packages from all over the place, and end up with some a lot less flexible than WireUI.

**Comment by judgej** [comment]
A lot of the underlying stuff described above are [on this page](https://livewire-wireui.com/docs/customization). I only just discoverewd it, right down the bottom of the menu.

**Comment by judgej** [comment]
Also just come across [component namespaces](https://laravel.com/docs/9.x/blade#manually-registering-components).

It should be possible for this package to register all components under a namespace, so `<x-input>` for example could be made available to the application as `<x-wireui::input>`. I'm not sure how easy this would be to implement, but could be a very useful feature.

Components can also be registered with a prefix:

    public function component($class, $alias = null, $prefix = '')
    {
        ...
    }

The prefix is added to the alias using a `-`. This this package could provide a global prefix (via config) for all components to namespace them. This is different to the namespacing described above. as the components would look more like `<x-wireui-input>`.

So many options: `<x-wireui::input>`, `<x-wireui-input>`, `<x-wireui.input>` - all prefixed by completely different mechanisms and approaches. I think the first is difficult without forcing ALL components to be active, not being able to switch any off by config. The last is the one I describe at the top of this thread. It is long-winded, but works. The one in the middle could be added to the package in a single config item - it would be a most useful feature.

Anybody else had this need, or is it just me?

**Reply by blockhunts** [reply]
I had the same problem as you too, old component and wireUI component name is similar.
Although being the lazy one, I just add [s] to the old component's name 😁
